stew/leb128.nim

```diff
@@ -0,0 +1,156 @@
+## Low-level little-endian base 128 variable length integer operators, as
+## described in https://en.wikipedia.org/wiki/LEB128 - up to 64 bits supported.
+##
+## Used in formats like DWARF, WASM - unsigned variant identical to protobuf/go
+## varint.
+##
+## This implementation contains low-level primitives suitable for building
+## more easy-to-use API.
+##
+## Exception/Defect free as of nim 1.2.
+
+{.push raises: [].}
+
+import
+  stew/bitops2
+
+const
+  # Given the truncated logarithm of a 64-bit number, how many bytes do we need
+  # to encode it?
+  lengths = block:
+    var v: array[64, int8]
+    for i in 0..<64:
+      v[i] = int8((i + 7) div 7)
+    v
+
+type
+  Leb128* = object
+    ## Type used to mark leb128 encoding helpers
+
+# log2trunc by definition never returns values >64, thus we can remove checks
+{.push checks: off.}
+func len*(T: type Leb128, x: SomeUnsignedInt): int8 =
+  ## Returns number of bytes required to encode integer ``x`` as leb128.
+  if x == 0: 1 # Always at least one byte!
+  else: lengths[log2trunc(x)]
+{.pop.}
+
+func maxLen*(T: type Leb128, I: type): int8 =
+  ## The maximum number of bytes needed to encode any value of type I
+  Leb128.len(I.high)
+
+type
+  Leb128Buf*[T: SomeUnsignedInt] = object
+    data*: array[maxLen(Leb128, T), byte] # len(data) <= 10
+    len*: int8 # >= 1 when holding valid protobuf
+
+  Leb128Error* = enum
+    incomplete = "End marker not found in given stream"
+    overflow = "Value would overflow"
+
+template write7(next: untyped) =
+  # write 7 bits of data
+  if v > type(v)(127):
+    result.data[result.len] = cast[byte](v) or 0x80'u8
```

Comment:
It is safer to use construction like this, because sometime you can get compiler error/warning.
```nim
result.data[result.len] = cast[byte](v and type(v)(0xFF)) or 0x80'u8
```